### ISSUE TITLE
Upload file

### ISSUE THREAD

**Issue by ibnufauzanalaula** [ufna/VaRest#262]
How to upload file (a csv file) along with other string data? For example, send data: name (string), id (string), and file (csv) to server

**Comment by ufna** [comment]
It's not possible for now, but I'll take a look for upcoming release.

**Comment by Minas314** [comment]
That would be great. I need it too. :)

**Comment by kasp1** [comment]
An ImageToByteArray function might do the job, as there is already a Base64Encode function for byte arrays in unreal. Base64 can be then uploaded as regular text and decoded server side. Not super efficient for frequent uploading but should do in most cases and would work with any byte arrays/files (not just images).

**Comment by tuncayson** [comment]
any progress for file uploading?

**Comment by Minas314** [comment]
This is how I upload files with VaRest:
![image](https://user-images.githubusercontent.com/10990243/137601263-ce77bb73-cb2f-4dad-b08b-5196fdcf3b95.png)

**Comment by IlliaImm** [comment]
@Minas314 Can you please tell what is under the hood in the node "Read bytes from file"? I wanted to try your method of uploading files, but I'm stack on converting file to byte array.

@kasp1 the same question, can you please guide a little about suggested ImageToByteArray function?

My case is - I make request to some API and the response is just one image in .png format. How can I process this response for further usage of this .png file.

**Comment by kasp1** [comment]
@IlliaImm If you develop the API as well then you would server the image as a Base64 encoded string, which you would download using a VaRest request and then use the Base64 decode function to turn it into a byte array, which you could then convert to a texture

If the image is presented as the image (already a byte array) you could use this unreal engine core function to turn it into a texture: https://docs.unrealengine.com/4.26/en-US/BlueprintAPI/Class/AsyncTaskDownloadImage/DownloadImage/

**Comment by IlliaImm** [comment]
@kasp1 Thank you for the response!

No, the API that I'm trying to communicate with is from some service provider, so I can't do anything with it's server side.

I tried that you recommended, but didn't succeeded in converting/decoding the response to smth that I can use. 
I attached screenshots to show what I get as a response and how I tried to decode it(i also tried without splitting the string - same result). And I'm not sure how I can use "DownloadImage" node, cause it requires simple URL. But my request consists not only of url, it also has a json object as a payload.

Hope you can help.

<details>
 <summary>Screenshots</summary>

![decode](https://user-images.githubusercontent.com/96781936/147744356-d1f1960a-5d0e-4f2c-9f1f-89c6e738dd5f.png)
![logmessage](https://user-images.githubusercontent.com/96781936/147744357-92dbcf4f-7214-43e1-85de-9c370761a397.png)
![screenpng](https://user-images.githubusercontent.com/96781936/147744358-52bf7bf2-8cd6-4a03-84e0-d955fc97ee90.png)


</details>

**Comment by kasp1** [comment]
@IlliaImm Nope, looks like you are trying to Base64 decode the actual image, which won't work because the image is not base64 encoded but raw image data already.

If I understand your case correctly, you need something like this:

![Screenshot_1](https://user-images.githubusercontent.com/11492515/147747624-493c0c4f-f7a3-46ef-b852-260b6550637e.png)

...or even better get the response _content as a byte array_ right away if that's possible?

Tip: You should be able to use the response Content-Type header to figure out if the content is png image (if the value of this header equals "image/png".

**Comment by IlliaImm** [comment]
@kasp1 Really appreciate your help. Sorry for bothering you many times.

Unfortunately, there is no option to get response _content as a byte array_. And unfortunately there is no unreal built-in function _StringToByteArray_. Also I checked if I can ask this API for some other types of data and the variants are only _image/png (object) or image/jpeg (object) or image/tiff (object)_. 

As you pointed in tip, I checked Content-Type and yes, it's image/png.
So for now I didn't make it work for me. I will look for some other options and if I will find a working solution I'll post it here.
I'm wondering if there is a possibility to just save response content as file in game directory and then just try to use this file?
If you have some other suggestions I would really like to try it.

Thank you again.

**Comment by kasp1** [comment]
Maybe this function could be wrapped into a custom blueprint function and used in blueprints https://docs.unrealengine.com/4.26/en-US/API/Runtime/Core/Containers/StringToBytes/

**Comment by IlliaImm** [comment]
I found solution that worked for me. I tried some different plugin that provides _Byte Data_ of the response. And this _Byte Data_ could be easily converted to a Texture by built-in function ImportBufferAsTexture2D. 

So it would be really nice to have in VaRest plugin a possibility to get Byte Data of the response and also a possibility to save it as a file in project directory.

**Comment by ssshammi** [comment]
HI could someone tell me how to get Read bytes from file i have a audio file from finish recording output and i just need to send this recorded output to the server via VaRest plugin. I can do this however i do not know hot to convert a file to base64 encoded string. 
![image](https://user-images.githubusercontent.com/2300969/162061375-2aa2aac0-6913-499a-ab08-b1393687f58b.png)
@Minas314 is this function "Read bytes from file" available in 4.27 or are you using a library or plugin?

**Comment by ssshammi** [comment]
![image](https://user-images.githubusercontent.com/2300969/162084869-ad32964f-0edc-485a-b0bb-eec621e3cb7f.png)
Hi is it possible to send form-data using VARest like use the request object and binary contents with files value.
![image](https://user-images.githubusercontent.com/2300969/162085012-9d85172f-0991-4f34-9a29-b071b193be50.png)

**Comment by chase03** [comment]
![Base64ToTexture2D](https://user-images.githubusercontent.com/26050994/180108402-e9dba31a-a2ac-4bbb-b850-40649f824053.jpg)

Hi! I make it work by using this way in UE5, Base64 to Texture2D

**Comment by ssshammi** [comment]
@chase03  Thanks i actually need to push to server, using multi-part form. And not read it. If you have been able to puch it let me know the data field is of type json and files.audioresponse is a binary file.

**Comment by IsmaelPozzo** [comment]
> This is how I upload files with VaRest: ![image](https://user-images.githubusercontent.com/10990243/137601263-ce77bb73-cb2f-4dad-b08b-5196fdcf3b95.png)
So far I don't understand where you get that function from called "Read Bytes from File", VArest doesn't have it, nor is there a code in this thread to do it in c++
